pkg/releaselock/controller.go

```diff
@@ -94,114 +117,232 @@ func NewLockReleaseController(client kubernetes.Interface, config *LockReleaseCo
-	nodeName, err := GKENodeNameFromConfigMap(cm)
+// TODO(b/374327452): interface rpc calls for mocking and create unit tests for handleCreateEvent and handleUpdateEvent
+func (c *LockReleaseController) handleCreateEvent(ctx context.Context, obj interface{}) error {
+	node := obj.(*corev1.Node)
+	start := time.Now()
+	cmName := ConfigMapNamePrefix + node.Name
+	cm, err := c.client.CoreV1().ConfigMaps(util.ManagedFilestoreCSINamespace).Get(ctx, cmName, metav1.GetOptions{})
+	duration := time.Since(start)
+	c.RecordKubeAPIMetrics(err, metrics.ConfigMapResourceType, metrics.GetOpType, metrics.ReconcilerOpSource, duration)
+
 	if err != nil {
-		klog.Errorf("Failed to get GKE node name from configmap %s/%s: %v", cm.Namespace, cm.Name, err)
+		klog.Errorf("Failed to get configmap in namespace %s: %v", util.ManagedFilestoreCSINamespace, err)
+		return nil
+	}
+	klog.Infof("Got configmap (%v) in namespace %s", cm, util.ManagedFilestoreCSINamespace)
+	data := cm.DeepCopy().Data
+	latestNode, err := c.client.CoreV1().Nodes().Get(ctx, node.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Failed to get node in namespace %v", err)
 		return err
 	}
 
-	node := nodes[nodeName]
-	data := cm.DeepCopy().Data
+	skippedEntry := false
+
 	for key, filestoreIP := range data {
 		_, _, _, _, gceInstanceID, gkeNodeInternalIP, err := ParseConfigMapKey(key)
 		if err != nil {
 			klog.Errorf("Failed to parse configmap key %s: %v", key, err)
 			continue
 		}
-		klog.V(6).Infof("Verifying GKE node %s with nodeId %s nodeInternalIP %s exists or not", nodeName, gceInstanceID, gkeNodeInternalIP)
-		nodeExists, err := c.verifyNodeExists(node, gceInstanceID, gkeNodeInternalIP)
+		klog.V(6).Infof("Verifying GKE node %s with nodeId %s nodeInternalIP %s exists or not", node.Name, gceInstanceID, gkeNodeInternalIP)
+		entryMatchesNode, err := c.verifyConfigMapEntry(node, gceInstanceID, gkeNodeInternalIP)
 		if err != nil {
-			klog.Errorf("Failed to verify GKE node %s with nodeId %s nodeInternalIP %s still exists: %v", nodeName, gceInstanceID, gkeNodeInternalIP, err)
+			klog.Errorf("Failed to verify GKE node %s with nodeId %s nodeInternalIP %s still exists: %v", node.Name, gceInstanceID, gkeNodeInternalIP, err)
+			skippedEntry = true
 			continue
 		}
-		if nodeExists {
-			klog.V(6).Infof("GKE node %s with nodeId %s nodeInternalIP %s still exists in API server, skip lock info reconciliation", nodeName, gceInstanceID, gkeNodeInternalIP)
+		if entryMatchesNode {
+			klog.V(6).Infof("GKE node %s with nodeId %s nodeInternalIP %s still exists in API server, skip lock info reconciliation", node.Name, gceInstanceID, gkeNodeInternalIP)
+			continue
+		}
+
+		// Try to match the latest node, to prevent incorrect releasing the lock in case of a lagging informer/watch
+		entryMatchesLatestNode, err := c.verifyConfigMapEntry(latestNode, gceInstanceID, gkeNodeInternalIP)
```

Comment:
We should query the API server to GET the latest node object

Reply:
we did GET the latest node in line 158

Reply:
will move line 158 to here